kanopy-platform/gateway-certificate-controller
# DEVOPS-3378 mutate external dns

### Description
This implements the following logic.
The istio gateway controller should be extended to control the values for the external-dns annotations:
* external-dns.alpha.kubernetes.io/hostname
* external-dns.alpha.kubernetes.io/target

The controller should accept a configuration that allows it to either discover the appropriate load balancer vanity name for the target value or accept a static value used for all target annotations. 

The controller should consult the namespace ingress-whitelist annotation to determine if the external-dns.alpha.kubernetes.io/target should be mutated.

If ingress-whitelist is not * then the target value should be mutated.

If ingress-whitelist is not * the external-dns.alpha.kubernetes.io/hostname annotation should always be removed if present. The external-dns gateway source always uses the host slice for dns values. 




I create a gateway that is:
```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: httpbin-gateway
  namespace: devops
  annotations:
    external-dns.alpha.kubernetes.io/hostname: "binhttp.gateway.example.com"
  labels:
    v1beta1.kanopy-platform.github.io/istio-cert-controller-inject-simple-credential-name: "true"
spec:
  servers:
    - port:
        number: 443
        name: https
        protocol: HTTPS
      hosts:
        - "devops/httpbin.gateway.example.com"
      tls:
        mode: SIMPLE
```

It becomes:
```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: httpbin-gateway
  namespace: devops
  labels:
    v1beta1.kanopy-platform.github.io/istio-cert-controller-inject-simple-credential-name: "true"
  annotations:
    external-dns.alpha.kubernetes.io/target: "ig-u.example.com"
spec:
  selector:
    istio: ingressgateway
    type: user
  servers:
    - port:
        number: 443
        name: https
        protocol: HTTPS
      hosts:
        - "devops/httpbin.gateway.example.com"
      tls:
        credentialName: devops-httpbin-gateway-https
        mode: SIMPLE
```

### Testing
Unit tests were added for confirming controller and mutating webhook behavior changes. 

Integration tests can be performed against minikube via skaffold:
1. skaffold dev
1. create a gateway with external-dns annotations for hostname and target
1. confirm that the hostname annotation is deleted and the target value is set to `overthere`

This version should function with an existing external-dns configuration with no change in behavior. That can be tested by running skaffold dev with the code changes and the example configuration from the tip of main.
